feat(cli): make url optional for login command (#10925)

elasticspoon opened the pull request:
Allow `coder login` to log into existing deployment if available and indicate URL used.

```sh
$ > go run ./cmd/coder/main.go login
Attempting to authenticate with saved URL: http://localhost:8080
Your browser has been opened to visit:

        http://localhost:8080/cli-auth

> Paste your token here: KtJjNBnP6D-EiYB7ADdUfOBBn66lVLXTy
> Welcome to Coder, admin! You're authenticated.
```

This PR also updates related error and help messages to indicate `coder login` is available as a command.

Fixes #10925
Fixes #9551

elasticspoon committed:
feat(cli): make url optional for login command (#10925)

Allow `coder login` to log into existing deployment if available.

Output to indicate saved URL was used.

Update help and error messages to indicate that `coder login` is
available as a command.

elasticspoon commented:
recheck

mafredri commented on cli/login.go:
```diff
@@ -136,18 +136,26 @@ func (r *RootCmd) login() *clibase.Cmd {
 				rawURL = inv.Args[0]
 			}
 
+			if url, err := r.createConfig().URL().Read(); rawURL == "" && err == nil {
```
Should we also check that `url != ""` just in case? Otherwise we may print the statement below which is misleading.

elasticspoon replied:
Potentially? I did it this way because if `url == ""` then that means the config/url exists but is empty so I think it makes more sense to print "Authing with config url: ". Maybe the url should be in quotes to indicate an empty url better.

I can see both sides here. Depends if you want to provide the extra info to the user or not.

elasticspoon replied:
Actually. By changing when the output happens the case of the misleading print statement has been addressed.

mafredri commented on cli/login.go:
```diff
@@ -136,18 +136,26 @@ func (r *RootCmd) login() *clibase.Cmd {
 			if len(inv.Args) == 0 {
 				rawURL = r.clientURL.String()
+				if rawURL != "" && rawURL == inv.Environ.Get(envURL) {
+					_, _ = fmt.Fprintf(inv.Stdout, "Attempting to authenticate with environment URL: %s\n", rawURL)
```
I wonder if we should verify the URL early, we're kind of giving the indication of success but then failing verbosely on bad input:

```
❯ CODER_URL=foo ./coder login
Attempting to authenticate with environment URL: foo
Encountered an error running "coder login"
Usage: coder login [flags] [<url>]
Error: Failed to check server "https://foo" for first user, is the URL correct and is coder accessible from your browser? Error - has initial user: Get "https://foo/api/v2/users/first": dial tcp: lookup foo on x.x.x.x:53: no such host
```

Side note: We also print `foo` only, not `https://foo` which is the default if no protocol is given.

Perhaps it might be ideal to delay the printing of messages so that we 1) note the source (`source = "config|environment|arg"`), then once we have successfully parsed the URL, we print the message noting the source.

elasticspoon replied:
I like that, I like that a lot. 

Additionally, if all the options are stored as source: `"config|environment|argument|flag"` if will also give the user a useful indication as to precedence. Ex: user calls `CODER_URL=foo coder login bar` it returns `Authing via arg: bar` to indicate arg > env var.

mafredri reviewed (commented):
Thank you for this PR! I have a couple of improvement suggestions but otherwise this looks really good.

elasticspoon committed:
feat(cli): output auth text on success only

It doesn't make sense to output Authing with ...
then fail. This moves the output later in the login
process to prevent that.

mafredri committed:
Merge branch 'main' into cli-use-config-url

mafredri committed:
make gen

mafredri reviewed (approved):
This turned out great, thanks for your contribution!
